### PR TITLE
Dall-E Image generation tool

### PR DESCRIPTION
**Pull request by davidfekke**
This PR includes a new integration for the OpenAI Dall-E image generation tool. This tool conforms to the `Tool` Langchain interface and supports both Dall-E 2 and Dall-E 3.

@davidfekke

### CODE REVIEW

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    quality (‘standard’ or ‘hd’): The quality of the image that will be generated. ‘hd’ creates images with finer details and greater consistency across the image. Defaults to ‘standard’.
+
+    Other parameters:
+    prompt (str): A text description of the desired image(s). The maximum length is 1000 characters. Required field.
+    n (int): The number of images to generate. Must be between 1 and 10. Defaults to 1. For dall-e-3, only n=1 is supported.
+    size (...): The size of the generated images. Must be one of 256x256, 512x512, or 1024x1024 for DALL·E-2 models. Must be one of 1024x1024, 1792x1024, or 1024x1792 for DALL·E-3 models.
+    response_format ('url' or 'b64_json'): The format in which the generated images are returned. Must be one of "url" or "b64_json". Defaults to "url".
+    user (str): A unique identifier representing your end-user, which will help OpenAI to monitor and detect abuse. Learn more.
+*/
+export interface DalleApiWrapperParams extends ToolParams {
```
Comment (bracesproul) [inline libs/langchain-community/src/tools/dalle_api_wrapper.ts:76]: Nice to have but not required: add jsdoc for each param with defaults if there is one. Eg:
```typescript
/**
 * The model to use
 * @default "dalle-3"
 */
modelName?: string;
```

File: libs/langchain-community/tools/dalle_api_warapper.cjs
Comment (bracesproul) [inline libs/langchain-community/tools/dalle_api_warapper.cjs:1]: these files shouldn't be committed

**Review by bracesproul — CHANGES_REQUESTED** [review]
Left some comments, thanks for pushing this up! Please re-request my review once they've been implemented!

File: examples/src/tools/dalle_image_generation.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+const params: DalleApiWrapperParams = {
+    prompt: "a painting of a cat",
+    n: 1,
+};
+
+const tool = new DalleApiWrapper(params);
+
+const imageArray = await tool.call("a painting of a cat");
+
+console.log(imageArray);
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply examples/src/tools/dalle_image_generation.ts:12]: I will add that image to the Docs.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    quality (‘standard’ or ‘hd’): The quality of the image that will be generated. ‘hd’ creates images with finer details and greater consistency across the image. Defaults to ‘standard’.
+
+    Other parameters:
+    prompt (str): A text description of the desired image(s). The maximum length is 1000 characters. Required field.
+    n (int): The number of images to generate. Must be between 1 and 10. Defaults to 1. For dall-e-3, only n=1 is supported.
+    size (...): The size of the generated images. Must be one of 256x256, 512x512, or 1024x1024 for DALL·E-2 models. Must be one of 1024x1024, 1792x1024, or 1024x1792 for DALL·E-3 models.
+    response_format ('url' or 'b64_json'): The format in which the generated images are returned. Must be one of "url" or "b64_json". Defaults to "url".
+    user (str): A unique identifier representing your end-user, which will help OpenAI to monitor and detect abuse. Learn more.
+*/
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:75]: I will change these to the proper types.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+export interface DalleApiWrapperParams extends ToolParams {
+    openAIApiKey?: string;
+
+    model?: DalleApiWrapperModelEnum;
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:79]: I will change this to a string.

File: examples/src/tools/dalle_image_generation.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { DalleApiWrapper, DalleApiWrapperParams } from '@langchain/community/tools/dalle_api_wrapper';
+
+const params: DalleApiWrapperParams = {
+    prompt: "a painting of a cat",
+    n: 1,
+};
+
+const tool = new DalleApiWrapper(params);
+
+const imageArray = await tool.call("a painting of a cat");
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply examples/src/tools/dalle_image_generation.ts:10]: I am changing the example to `.invoke`.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import { type ClientOptions, OpenAI as OpenAIClient } from "openai";
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:2]: I will add these dependencies to the project.json for the `langchain-community` lib and the `langchain.config.js`.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    name = "dalle_api_wrapper";
+
+    description = `A OpenAI Dall-E tool. useful for generating images with Open AIs Dall-E 2 or 3 API.`;
+
+    organization?: string;
+
+    protected openAIApiKey?: string;
+    protected client: OpenAIClient;
+    private clientConfig: ClientOptions;
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:114]: I need to keep the `client`, but I am deleting the `openAIApiKey` and the `clientConfig` from the class properties.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    quality?: qualityEnum;
+
+    prompt: string;
+
+    n?: number;
+
+    size?: sizeEnum;
+
+    response_format?: responseFormatEnum;
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:91]: I am making this camelCase.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import { type ClientOptions, OpenAI as OpenAIClient } from "openai";
+import { Tool, ToolParams } from "@langchain/core/tools";
+
+/*
+    An enumerator for the Dall-E model to use with the OpenAI.
+    Dall-E 2 and 3 API
+    https://beta.openai.com/docs/api-reference/images/generate
+ */
```
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:9]: I am adding JSDocs for all of these properties.

File: libs/langchain-community/src/tools/dalle_api_wrapper.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidfekke) [reply libs/langchain-community/src/tools/dalle_api_wrapper.ts:79]: I am also changing the `model` to `modelName`.

**Review by davidfekke — COMMENTED** [review]
I have made changes based on suggestions from @bracesproul.

File: .gitignore [hunk 10]
```diff
@@ -43,3 +43,5 @@ langchain/api_refs_docs_build/dist/**/*
 .docusaurus/
 docs/build/
 docs/api_refs/typedoc.json
+
+.tool-versions
```
Comment (bracesproul) [inline .gitignore:47]: Why was this added?

Comment (davidfekke) [reply .gitignore:47]: `.tool-versions` is a file used by the `asdf` tool to specify which versions of different runtimes you want to use. Many Node.js developers use this instead of NVM to pick which version of Node.js they are using. You can specify a specific version of Node.js for a local workspace or folder.

Comment (bracesproul) [reply .gitignore:47]: Ah good to know. Thank you!

File: libs/langchain-openai/src/tools/dalle.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+ * A tool for generating images with Open AIs Dall-E 2 or 3 API.
+ */
+export class DallEAPIWrapper extends Tool {
+  static lc_name() {
+    return "DallEAPIWrapper";
+  }
+
+  name = "dalle_api_wrapper";
+
+  description = `Dall-E tool. Useful for generating images with Open AIs Dall-E API.`;
```
Comment (bracesproul) [inline libs/langchain-openai/src/tools/dalle.ts:73]:
```suggestion
  description = "A wrapper around OpenAI DALL-E API. Useful for when you need to generate images from a text description. Input should be an image description.";
```

Comment (bracesproul) [reply libs/langchain-openai/src/tools/dalle.ts:73]: port description 1:1 from py^

**Review by bracesproul — APPROVED** [review]
thanks for adding this!

**Review by davidfekke — COMMENTED** [review]
Perfect!

**Comment by jacoblee93** [comment]
Thank you!

**Comment by davidfekke** [comment]
Thank you for helping me get this added to LangChain js.
